src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/SpellCheck/DocumentSpellCheckEndpoint.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.SpellCheck;
+
+[LanguageServerEndpoint(VSInternalMethods.TextDocumentSpellCheckableRangesName)]
+internal sealed class DocumentSpellCheckEndpoint : IRazorRequestHandler<VSInternalDocumentSpellCheckableParams, VSInternalSpellCheckableRangeReport[]>, IRegistrationExtension
+{
+    private readonly IRazorDocumentMappingService _documentMappingService;
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+    private readonly ClientNotifierServiceBase _languageServer;
+
+    public DocumentSpellCheckEndpoint(
+        IRazorDocumentMappingService documentMappingService,
+        LanguageServerFeatureOptions languageServerFeatureOptions,
+        ClientNotifierServiceBase languageServer)
+    {
+        _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+        _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public RegistrationExtensionResult GetRegistration(VSInternalClientCapabilities clientCapabilities)
+    {
+        const string ServerCapability = "_vs_spellCheckingProvider";
+
+        return new RegistrationExtensionResult(ServerCapability, true);
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalDocumentSpellCheckableParams request)
+    {
+        if (request.TextDocument is null)
+        {
+            throw new ArgumentNullException(nameof(request.TextDocument));
+        }
+
+        return request.TextDocument;
+    }
+
+    public async Task<VSInternalSpellCheckableRangeReport[]> HandleRequestAsync(VSInternalDocumentSpellCheckableParams request, RazorRequestContext requestContext, CancellationToken cancellationToken)
```

Comment:
This doesn't at the moment, and its a little problematic. There is a limitation in the platform that we can't stream results from our custom message target. We could in theory allow Roslyn to stream the results, and collect the chunks in that target, but we'd have to send all of Roslyns results over in one go. Similarly we need to have all of Roslyns results here in one chunk, because we need to combine them with the Razor results, and being a delta based API, everything has to be in order.

We could in theory, once we've done all of the work, send the results back to the client in chunks from here, but I'm not sure that alone is worth it.